### PR TITLE
Kernel+LibC: Add getrusage()

### LINKED ISSUE #1140: LibC: Implement getrusage function
**https://linux.die.net/man/3/getrusage**

### PR DESCRIPTION
**Pull request by LucasChollet**
Closes #1140

Add the getrusage() function for POSIX compliance

### CODE REVIEW

File: Userland/Utilities/CMakeLists.txt [hunk 1]
```diff
@@ -7,9 +7,9 @@ list(APPEND REQUIRED_TARGETS
     touch tr true umount uname uniq uptime w wc which whoami xargs yes less
 )
 list(APPEND RECOMMENDED_TARGETS
-    adjtime aplay abench asctl bt checksum chres cksum copy fortune gunzip gzip init keymap lsirq lsof lspci man mknod mktemp
+    tester adjtime aplay abench asctl bt checksum chres cksum copy fortune gunzip gzip init keymap lsirq lsof lspci man mknod mktemp
```
Comment (linusg) [inline Userland/Utilities/CMakeLists.txt:10]: ?

File: Kernel/Syscalls/resource.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Copyright (c) 2022, Lucas Chollet <lucas.chollet@free.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Process.h>
+
+namespace Kernel {
+
+// https://pubs.opengroup.org/onlinepubs/009696699/functions/getrusage.html
```
Comment (linusg) [inline Kernel/Syscalls/resource.cpp:11]: This comment should be on the LibC function, not the syscall

File: Userland/Libraries/LibC/ulimit.cpp [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
 #include <assert.h>
 #include <sys/resource.h>
+#include <syscall.h>
 #include <ulimit.h>
 
 extern "C" {
 
 long ulimit([[maybe_unused]] int cmd, [[maybe_unused]] long newlimit)
 {
-    dbgln("FIXME: Implement getrusage()");
```
Comment (linusg) [inline Userland/Libraries/LibC/ulimit.cpp:16]: Why is the dbgln being removed if the function still crashes? Please update the message instead.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Kernel/API/POSIX/sys/resource.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct rusage {
+    struct timeval ru_utime {
+    };
+    struct timeval ru_stime {
+    };
+    long ru_maxrss {};
```
Comment (bgianfo) [inline Kernel/API/POSIX/sys/resource.h:21]: This isn't valid C, so if a port tries to include this from LibC it won't compile?

File: Userland/Utilities/CMakeLists.txt (hunk 1 above)
Comment (bgianfo) [inline Userland/Utilities/CMakeLists.txt]: Included by mistake?

File: Kernel/Syscalls/resource.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <Kernel/Process.h>
+
+namespace Kernel {
+
+ErrorOr<FlatPtr> Process::sys$getrusage(int who, Userspace<rusage*> user_usage)
+{
+    VERIFY_PROCESS_BIG_LOCK_ACQUIRED(this)
+
+    rusage usage;
```
Comment (bgianfo) [inline Kernel/Syscalls/resource.cpp:15]: You will need to zero initialize this after you fix the other comment about having a C++ constructor in a C header.
```suggestion
    rusage usage {};
```

**Review by bgianfo — CHANGES_REQUESTED** [review]

File: Kernel/API/POSIX/sys/resource.h (hunk 4 above)
Comment (LucasChollet) [reply Kernel/API/POSIX/sys/resource.h]: I wondered for a minute but thought it was ok as it compiles

File: Kernel/Syscalls/resource.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include <Kernel/Process.h>
+
+namespace Kernel {
+
+ErrorOr<FlatPtr> Process::sys$getrusage(int who, Userspace<rusage*> user_usage)
+{
+    VERIFY_PROCESS_BIG_LOCK_ACQUIRED(this)
+
```
Comment (bgianfo) [inline Kernel/Syscalls/resource.cpp:14]: We should figure out which pledge we should lock this down under, as I don't think we want `getrusage(..)` to be usable by a restricted program.

It seems like we might want to introduce a new pledge just for this. 

@awesomekling Any thoughts?

**Review by bgianfo — APPROVED** [review]

File: Kernel/Syscalls/resource.cpp (hunk 6 above)
Comment (louisdem) [reply Kernel/Syscalls/resource.cpp:14]: Hi @awesomekling! :^) Can you share your thoughts on this?

File: AK/Time.cpp [hunk 7]
```diff
@@ -38,6 +38,15 @@ unsigned day_of_week(int year, unsigned month, int day)
     return (year + year / 4 - year / 100 + year / 400 + seek_table[month - 1] + day) % 7;
 }
 
+Time Time::from_ticks(clock_t ticks, time_t ticks_per_second)
+{
+    auto secs = ticks % ticks_per_second;
+
+    i32 nsecs = 1'000'000'000 * (ticks - (ticks_per_second * secs)) / ticks_per_second;
```
Comment (awesomekling) [inline AK/Time.cpp:45]: This looks extremely likely to overflow when using `i32`. Any reason to not use `i64` here?

Comment (LucasChollet) [reply AK/Time.cpp:45]: AFAICT the computation needs to be done on 64 bits, but the result will always fit on 32 bits.
I'm not entirely sure, but this should be good with integer promotions.
```ticks - (ticks_per_second * secs)``` will always be one power of then smaller than ```ticks_per_second``` so ```ticks - (ticks_per_second * secs) / ticks_per_second``` will be lower than 1. By the end i32 is smaller than ```1e9``` i.e. doesn't overflow.

**Pull request merged**